statistics/feedback.go

```diff
@@ -177,3 +175,321 @@ func (q *QueryFeedback) Update(startKey kv.Key, counts []int64) {
+		highVal = convertBytesToScalar(high.GetBytes()[common:])
+	} else {
+		lowVal, midVal, highVal = float64(low.GetInt64()), float64(mid.GetInt64()), float64(high.GetInt64())
+	}
+	// If we choose to merge, err is the absolute estimate error for the previous bucket.
+	err := calcFraction(lowVal, highVal, midVal)*(preCount+count) - preCount
+	return bucketScore{id, math.Abs(err / (preCount + count))}
+}
+
+// defaultBucketCount is the number of buckets a column histogram has.
+var defaultBucketCount = 256
+
+func mergeBuckets(bkts []bucket, isNewBuckets []bool, totalCount float64) []bucket {
+	mergeCount := len(bkts) - defaultBucketCount
+	if mergeCount <= 0 {
+		return bkts
+	}
+	bs := make(bucketScores, 0, len(bkts))
+	for i := 1; i < len(bkts); i++ {
+		// Do not merge the newly created buckets.
+		if !isNewBuckets[i] && !isNewBuckets[i-1] {
+			bs = append(bs, getBucketScore(bkts, totalCount, i))
+		}
+	}
+	sort.Sort(bs)
+	ids := make([]int, 0, mergeCount)
+	for i := 0; i < mergeCount; i++ {
+		ids = append(ids, bs[i].id)
+	}
+	sort.Ints(ids)
+	idCursor, bktCursor := 0, 0
+	for i := range bkts {
+		// Merge this bucket with last one.
+		if idCursor < mergeCount && ids[idCursor] == i {
+			bkts[bktCursor-1].upper = bkts[i].upper
+			bkts[bktCursor-1].count += bkts[i].count
+			bkts[bktCursor-1].repeat = bkts[i].repeat
+			idCursor++
+		} else {
+			bkts[bktCursor] = bkts[i]
+			bktCursor++
+		}
+	}
+	bkts = bkts[:bktCursor]
+	return bkts
+}
+
+func splitBuckets(h *Histogram, feedbacks []*QueryFeedback) ([]bucket, []bool, int64) {
+	bktFB, fbNum := buildBucketFeedback(h, feedbacks)
+	counts := make([]int64, 0, h.Len())
+	for i := 0; i < h.Len(); i++ {
+		bkt, ok := bktFB[i]
+		if !ok {
+			counts = append(counts, h.bucketCount(i))
+		} else {
+			counts = append(counts, bkt.getBucketCount())
+		}
+	}
+	totCount := int64(0)
+	for _, count := range counts {
+		totCount += count
+	}
+	buckets := make([]bucket, 0, h.Len())
+	isNewBuckets := make([]bool, 0, h.Len())
+	splitCount := getSplitCount(fbNum)
+	for i := 0; i < h.Len(); i++ {
+		bkt, ok := bktFB[i]
+		// No feedback, just use the original one.
+		if !ok {
+			buckets = append(buckets, bucket{h.GetLower(i), h.GetUpper(i), counts[i], h.Buckets[i].Repeat})
+			isNewBuckets = append(isNewBuckets, false)
+			continue
+		}
+		bkts := bkt.splitBucket(splitCount*len(bkt.feedback)/fbNum, float64(totCount), float64(counts[i]))
+		buckets = append(buckets, bkts...)
+		if len(bkts) == 1 {
+			isNewBuckets = append(isNewBuckets, false)
+		} else {
+			for i := 0; i < len(bkts); i++ {
+				isNewBuckets = append(isNewBuckets, true)
+			}
+		}
+	}
+	return buckets, isNewBuckets, totCount
+}
+
+// UpdateHistogram updates the histogram according buckets.
+func UpdateHistogram(h *Histogram, feedbacks []*QueryFeedback) *Histogram {
+	buckets, isNewBuckets, totalCount := splitBuckets(h, feedbacks)
+	buckets = mergeBuckets(buckets, isNewBuckets, float64(totalCount))
```

Comment:
Will the number of the buckets exceed 256 after merge?

Reply:
No, because we will split at most 10 buckets, so we only need to find 10 buckets to merge which is not hard.